moor_generator/lib/src/analyzer/sql_queries/query_analyzer.dart

```diff
@@ -28,10 +30,24 @@ abstract class BaseAnalyzer {
     if (_engine == null) {
       _engine = step.task.session.spawnEngine();
       tables.map(mapper.extractStructure).forEach(_engine.registerTable);
+      resolveViews();
+      views.map(mapper.extractView).forEach(_engine.registerView);
     }
     return _engine;
   }
 
+  /// Parses the view and adds columns to its resolved columns.
+  @protected
+  void resolveViews() {
+    for (final view in views) {
+      final ctx = _engine.analyzeNode(
```

Comment:
Do we have to do that analysis here? This class is mainly used for analyzing queries and the body of triggers.

Can we analyze and resolve views earlier, say in `MoorParser.parseAndAnalyze`?

Reply:
Initially I wanted to have it there, but don't we need all views and tables to be registered in the engine prior to analyzing? Since the body of a view is a query I thought that here is a better place, this way if we have some table definitions in another file and we're importing them, we can still create views.

Reply:
I also added another test that fails the current implementation, making a view from another view, so I need to find a way to resolve views topologically. Queries don't suffer from this because they are created after all tables and import statements.

Reply:
> but don't we need all views and tables to be registered in the engine prior to analyzing

Right! Now I remember why I was hesitating to implement views for so long :D

Views referencing other views make the whole situation even harder for sure. If you want to we can ignore that case for a moment (we can add an error message if a view references another view). It's a hard problem and it doesn't map well to the current way moor resolves files.

We currently have a two-step procedure to analyze moor files, and we guarantee that all dependencies have step 1 completed when we enter step 2 of any moor file:

1. Parsing: Parse the moor file and resolve local information (e.g. the table schema) (`ParseMoorStep`)
2. Analysis: Resolve queries and foreign references that need information from other files (`AnalyzeMoorStep`)

We could add another step between them to resolve views, and then ensure that views of dependents have been resolved once we get to step 3. I'm not sure if there's an easier solution that I'm overlooking.

Reply:
I guess this can be done in the `AnalyzeMoorStep`:

Currently we're using `SqlAnalyzer` prior to the `EntityHandler`, I think if we swap them to first get the references for views, and then have something like `ViewAnalyzer.resolve()` which resolves the views before calling the `SqlAnalyzer.parse()`, that would solve this problem.